Refactoring last tag from Link Header

adusumillipraveen opened the pull request:
- Refactoring code to get lastTag.
- Current logic fails when using orderBy
- See https://github.com/Azure/acr-cli/pull/72#issuecomment-664201313

timja reviewed (approved):
Thanks Praveen!

Easier to review PR with whitespace diff disabled:
https://github.com/Azure/acr-cli/pull/73/files?w=1

Wwwsylvia commented on cmd/acr/purge.go:
```diff
@@ -274,14 +279,39 @@ func getTagsToDelete(ctx context.Context,
+				for _, queryParam := range queryParams {
+					lastParam := strings.Split(queryParam, "last=")
+					if len(lastParam) > 1 {
+						newLastTag = lastParam[1]
```
Should we break the loop if we've found the last tag?

Wwwsylvia commented on cmd/acr/purge.go:
```diff
@@ -274,14 +279,39 @@ func getTagsToDelete(ctx context.Context,
 	return nil, "", nil
 }
 
+func getLastTagFromResponse(resultTags *acr.RepositoryTagsType) string {
```
FYI, if the request tags are ordered by time, the returned `last` parameter may contains "&" character, which may cause unexpected issues during URL encoding and decoding.

Wwwsylvia commented on cmd/acr/purge_test.go:
```diff
@@ -606,6 +637,30 @@ var (
+		Response: autorest.Response{
+			Response: &http.Response{
+				StatusCode: 200,
+				Header:     http.Header{linkHeader: {"/acr/v1/&testRepo/_tags?last=latest&n=3&orderby=timedesc"}},
```
Should we mock the real link header format here? Like `</acr/v1/&testRepo/_tags?last=latest&n=3&orderby=timedesc>; rel="next"`

Besides, can we have a few more test cases to cover different possible queries? Such as `</acr/v1/library/ubuntu/_tags?last=123%26latest&n=100&orderby=timedesc>; rel="next"`, `</acr/v1/library/ubuntu/_tags?orderby=&n=100&last=123%26latest>; rel="next"`, etc.

(`%26` is `&`)

adusumillipraveen replied:
Added tests for this

Wwwsylvia commented:
Hi @adusumillipraveen , thanks for the PR! 🌷 I left some comments and hope them help.
I also tested it on my local machine and caught some bugs...
If we set the `orderby` to `timedesc`, we will find the returned `last` parameter contains `&` character, which will be encoded as `%26`. However, if we take the encoded tag as a parameter of next request, the `%26` will be encoded to `%2526`, which will leads to unexpected result. 

![image](https://user-images.githubusercontent.com/8113721/91550983-064a0a00-e95c-11ea-904f-508a68175104.png)

Wwwsylvia commented on cmd/acr/purge.go:
```diff
@@ -274,14 +279,39 @@ func getTagsToDelete(ctx context.Context,
+			if len(queryString) > 1 {
+				stripRel := strings.Split(queryString[1], ";")
+
+				queryParams := strings.Split(stripRel[0], "&")
```
I tested it on my local machine and caught some bugs...
If we set the `orderby` to `timedesc`, we will find the returned `last` parameter contains `&` character, which will be encoded as `%26`. However, if we take the encoded tag as a parameter of next request, the `%26` will be encoded to `%2526` subsequently, which will leads to unexpected result.

My point is, we have to be careful of the manual URL parsing.

![image](https://user-images.githubusercontent.com/8113721/91550983-064a0a00-e95c-11ea-904f-508a68175104.png)
The code in snapshot is at https://github.com/Azure/acr-cli/blob/3cd135bf0f05bfa49e7f78d01f4298269d0a8066/acr/client.go#L1693-L1695

adusumillipraveen commented:
@Wwwsylvia Thanks for your quick review , was quite busy last week and had change only today to pick this up. I have addressed most of the comments you have raised . Can you please let me know if its missing any more scenarios ? 

Also, I am thinking https://github.com/Azure/acr-cli/pull/73/files#diff-bd8183c84337996183d0bb9280f6e3efR592 is wrong and it needs to be 123%26latest ?Will try to test it on a repo and update my findings here .

adusumillipraveen commented:
@Wwwsylvia  I have tested some more scenarios by merging to https://github.com/Azure/acr-cli/pull/72 and I can see it working fine with '&' while using orderby. Please let me know if u see any issues. This is the branch if you want to give a try https://github.com/hmcts/acr-cli/tree/add-keep-flag.

adusumillipraveen commented:
@Wwwsylvia Have you had a chance to look at the latest changes ?

Wwwsylvia commented:
> @Wwwsylvia Have you had a chance to look at the latest changes ?

Sorry, been busy these days... Will find a chance to check this and get back to you.

adusumillipraveen commented:
> Sorry, been busy these days... Will find a chance to check this and get back to you.

Hey @Wwwsylvia was wondering if you had a chance to look at it ?

Wwwsylvia commented on cmd/acr/purge.go:
```diff
@@ -400,27 +425,30 @@ func dryRunPurge(ctx context.Context, acrClient api.AcrCLIClientInterface, login
-		tagsToDelete, lastTag, err = getTagsToDelete(ctx, acrClient, repoName, regex, timeToCompare, lastTag)
+	for {
+		tagsToDelete, newLastTag, err := getTagsToDelete(ctx, acrClient, repoName, regex, timeToCompare, lastTag)
+		lastTag = newLastTag
```
Can we check the error first? Same as we did in purgeTags()
https://github.com/Azure/acr-cli/pull/73/files#diff-6a5ebae231898bd993422487862b3818R179

adusumillipraveen replied:
Missed this one 🤦

Wwwsylvia commented on cmd/acr/purge_test.go:
```diff
@@ -606,6 +650,68 @@ var (
+		Response: autorest.Response{
+			Response: &http.Response{
+				StatusCode: 200,
+				Header:     http.Header{linkHeader: {"/acr/v1/&testRepo/_tags?last=latest&n=3&orderby=timedesc rel=\"next\""}},
```
Shouldn't we test the format `</acr/v1/&testRepo/_tags?last=latest&n=3&orderby=timedesc>; rel="next"` ?

Wwwsylvia replied:
@adusumillipraveen Any ideas on this? and other similar ones.

adusumillipraveen replied:
I changed it and it breaks one of my test , but it works on actual repos i am trying to purge. I might be missing something in there, will spend some time to see if anything else is trimming them in case of non stubbed code.

Wwwsylvia replied:
I'm wondering if there is parsing problem when the link is like `</acr/v1/&testRepo/_tags?n=3&orderby=timedesc&</acr/v1/&testRepo/_tags?last=latest&n=3&orderby=timedesc>; rel="next">; rel="next"`

adusumillipraveen replied:
Yes, I am not sure of that scenario is really need though. But i have pushed a fix to address that. Please let me know if that looks ok.

Wwwsylvia commented on cmd/acr/purge_test.go:
```diff
@@ -606,6 +650,68 @@ var (
+		},
+	}
+
+	OneTagResultQueryEndingWithLast = &acr.RepositoryTagsType{
```
What is the difference between `OneTagResultQueryEndingWithLast` and `OneTagResultWithAmpersand`? They look the same... Do you mean `</acr/v1/library/ubuntu/_tags?orderby=&n=100&last=123%26latest>; rel="next"` here?

adusumillipraveen replied:
yes, i definitely meant that.

Wwwsylvia reviewed (changes requested):
Hi @adusumillipraveen, sorry for the delay! I've reviewed and left a few more comments. The PR overall looks good, just some test cases are a little bit confusing.
